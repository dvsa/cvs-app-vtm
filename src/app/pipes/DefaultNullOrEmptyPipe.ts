import { Pipe, PipeTransform } from '@angular/core';

@Pipe({ name: 'DefaultNullOrEmpty' })
export class DefaultNullOrEmpty implements PipeTransform {
  transform(value: any): any {
    switch (typeof value) {
      case 'string':
        if (value.trim().length > 0) {
          return value;
        } else {
          return '-';
        }
        break;
      case 'boolean':
        return value ? 'Yes' : 'No';
        break;
      default:
        return value == null ? '-' : value;
    }
  }
}
