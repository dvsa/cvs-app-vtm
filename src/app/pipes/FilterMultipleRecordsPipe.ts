import { Pipe, PipeTransform } from '@angular/core';
import { TechRecord } from '@app/models/tech-record.model';
import { VehicleTechRecordModel } from '@app/models/vehicle-tech-record.model';

@Pipe({ name: 'FilterMultipleRecords' })
export class FilterMultipleRecordsPipe implements PipeTransform {
  transform(multipleRecordsList: VehicleTechRecordModel[]): VehicleTechRecordModel[] {
    const vehicleTechnicalRecords = [...multipleRecordsList];

    vehicleTechnicalRecords.sort((a, b) => {
      const recordA = this.getRecord(a);
      const recordB = this.getRecord(b);

      if (recordA.make !== recordB.make) {
        return recordA.make.localeCompare(recordB.make);
      } else {
        return recordB.manufactureYear - recordA.manufactureYear;
      }
    });
    return vehicleTechnicalRecords;
  }

  getRecord(vehicleTechRecord: VehicleTechRecordModel): TechRecord {
    let record: TechRecord;

    record = vehicleTechRecord.techRecord.find(
      (tRec: TechRecord) => tRec.statusCode === 'current'
    );
    if (record !== undefined) {
      return record;
    }

    record = vehicleTechRecord.techRecord.find(
      (tRec: TechRecord) => tRec.statusCode === 'provisional'
    );
    if (record !== undefined) {
      return record;
    }

    record = vehicleTechRecord.techRecord
      .filter((tRec: TechRecord) => tRec.statusCode === 'archived')
      .sort((a, b) => new Date(b.createdAt).getTime() - new Date(a.createdAt).getTime())[0];

    return record;
  }
}
