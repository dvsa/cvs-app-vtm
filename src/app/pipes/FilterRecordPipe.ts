import { Pipe, PipeTransform } from '@angular/core';
@Pipe({name: 'FilterRecord'})
export class FilterRecordPipe implements PipeTransform {

  transform(techRecordList: any): any {
    if (techRecordList) {
      let records = techRecordList.find((record: any) => record.statusCode === 'current');
      if (records !== undefined) return records;

      records = techRecordList.find((record: any) => record.statusCode === 'provisional');
      if (records !== undefined) return records;

      records = techRecordList.filter((record: any) => record.statusCode === 'archived');
      records.sort((a,b)=> new Date (b.createdAt).getTime() - new Date (a.createdAt).getTime() );
      if (records !== undefined) return records[0];
    }
  }
}
