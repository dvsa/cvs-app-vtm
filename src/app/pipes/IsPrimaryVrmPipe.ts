import { Pipe, PipeTransform } from '@angular/core';
@Pipe({name: 'IsPrimaryVrm'})
export class IsPrimaryVrmPipe implements PipeTransform {

  transform(vrmsList: any, isPrimary: boolean): any[] {
    if (vrmsList) {
      return vrmsList.filter((listing: any) => listing.isPrimary === isPrimary);
    }
  }
}
