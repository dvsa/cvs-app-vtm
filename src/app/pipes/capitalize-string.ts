import { Pipe, PipeTransform } from '@angular/core';

@Pipe({ name: 'CapitalizeString' })
export class CapitalizeString implements PipeTransform {
  transform(string: string): string {
    return string.length > 0
      ? string.charAt(0).toUpperCase() + string.substr(1).toLowerCase()
      : '';
  }
}
