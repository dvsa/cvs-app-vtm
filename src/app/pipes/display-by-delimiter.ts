import { Pipe, PipeTransform } from '@angular/core';

@Pipe({ name: 'DisplayByDelimiter' })
export class DisplayByDelimiter implements PipeTransform {
  transform(strings: (number | string)[], delimiter: string): string {
    const stringsToJoin = strings.length > 1 ? strings.filter((item) => item) : [''];
    return stringsToJoin.join(delimiter);
  }
}
