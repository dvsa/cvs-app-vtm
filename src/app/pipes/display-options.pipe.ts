import { Pipe, PipeTransform } from '@angular/core';
import { SelectOption } from './../models/select-option';

/**
 *
 * @usageNotes
 *
 * Converts a string to a dev friendly identifiable object
 *
 */
@Pipe({
  name: 'displayOptions'
})
export class DisplayOptionsPipe implements PipeTransform {
  transform(strOptions: string[], selectedOptions: string[] = []): SelectOption[] {
    return strOptions.map((option, index) => {
      return {
        id: index,
        name: option.trim(),
        selected: selectedOptions.includes(option)
      } as SelectOption;
    });
  }
}
