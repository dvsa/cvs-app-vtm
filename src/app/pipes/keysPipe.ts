import {Pipe, PipeTransform} from '@angular/core';

@Pipe({name: 'keys'})
export class KeysPipe implements PipeTransform {
  transform(value: { [x: string]: any; }, args?: string[]): any {
    const keys = [];

    console.log(value);

    let incr = 0;
    for (const key in value) {
      keys.push({ key: value[key].content, value: value[key].value, index: incr });
      incr++;
    }

    console.log(keys);

    return keys;
  }
}

