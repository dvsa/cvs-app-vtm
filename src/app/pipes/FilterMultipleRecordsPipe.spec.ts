import { FilterMultipleRecordsPipe } from '@app/pipes/FilterMultipleRecordsPipe';

const vehicleTechRecordList = [
  {
    techRecord: [
      {
        vin: '1234568',
        statusCode: 'current',
        make: 'Isuzu',
        manufactureYear: 2018
      }
    ]
  },
  {
    techRecord: [
      {
        vin: '1234567',
        statusCode: 'provisional',
        make: 'AIsuzu',
        manufactureYear: 2019
      },
    ]
  },
  {
    techRecord: [
      {
        vin: '1234569',
        statusCode: 'archived',
        createdAt: '2019-06-25T10:26:54.903Z',
        make: 'AIsuzu',
        manufactureYear: 2020
      },
      {
        vin: '1234566',
        statusCode: 'archived',
        createdAt: '2019-06-26T08:26:54.903Z'
      }
    ]
  }
];

const orderedVehicleTechRecordList = [
  {
    techRecord: [
      {
        vin: '1234568',
        statusCode: 'current',
        make: 'Isuzu',
        manufactureYear: 2018
      }
    ]
  },
  {
    techRecord: [
      {
        vin: '1234567',
        statusCode: 'provisional',
        make: 'AIsuzu',
        manufactureYear: 2019
      },
    ]
  },
];



describe('FilterMultipleRecordsPipe', () => {
  const pipe = new FilterMultipleRecordsPipe();

  it('create an instance', () => {
    expect(pipe).toBeTruthy();
  });

  it('should return undefined if technical record list is null', () => {
    const vTechRecordList = pipe.transform(null);
    expect(vTechRecordList).toBeUndefined();
  });

  it('should return a vehicle tech record list sorted by make or manufacturer Year', () => {
    const vTechRecordList = pipe.transform(orderedVehicleTechRecordList);
    expect(vTechRecordList).toBeDefined();
    expect(vTechRecordList).toEqual(orderedVehicleTechRecordList);
  });

  it('if it CURRENT techRec it should set the technical record having status CURRENT for sorting', () => {
    const vTechRecordList = pipe.transform(vehicleTechRecordList);
    expect(vTechRecordList).toBeDefined();
    expect(vTechRecordList[0].techRecord[1].statusCode).toEqual('provisional');
  });

  it('if it hasn;t CURRENT techRec it should set the technical record having status PROVISIONAL for sorting', () => {
    const vTechRecordList = pipe.transform(vehicleTechRecordList);
    expect(vTechRecordList).toBeDefined();
    expect(vTechRecordList[1].techRecord[1].statusCode).toEqual('current');
  });

  it('if it hasn;t CURRENT/PROVISIONAL techRec it should set the technical record having status ARCHIVED for sorting', () => {
    const vTechRecordList = pipe.transform(vehicleTechRecordList);
    expect(vTechRecordList).toBeDefined();
    expect(vTechRecordList[2].techRecord[1].statusCode).toEqual('archived');
  });

});
