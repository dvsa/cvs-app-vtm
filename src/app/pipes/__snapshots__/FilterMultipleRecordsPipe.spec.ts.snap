// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilterMultipleRecordsPipe should return a vehicle tech record list sorted by make or manufacturer Year 1`] = `
Array [
  Object {
    "techRecord": Array [
      Array [
        Object {
          "make": "AIsuzu",
          "manufactureYear": 2019,
          "statusCode": "provisional",
        },
      ],
      Object {
        "make": "AIsuzu",
        "manufactureYear": 2019,
        "statusCode": "provisional",
      },
    ],
  },
  Object {
    "techRecord": Array [
      Array [
        Object {
          "make": "Isuzu",
          "manufactureYear": 2018,
          "statusCode": "current",
        },
      ],
      Object {
        "make": "Isuzu",
        "manufactureYear": 2018,
        "statusCode": "current",
      },
    ],
  },
  Object {
    "techRecord": Array [
      Array [
        Object {
          "createdAt": "2019-06-25T10:26:54.903Z",
          "make": "AIsuzu",
          "manufactureYear": 2020,
          "statusCode": "archived",
        },
        Object {
          "createdAt": "2019-06-26T08:26:54.903Z",
          "statusCode": "archived",
        },
      ],
      Object {
        "createdAt": "2019-06-26T08:26:54.903Z",
        "statusCode": "archived",
      },
    ],
  },
]
`;
